### PR TITLE
[SPARK-24762][SQL] Enable Option of Product encoders

### PR DESCRIPTION
**Pull request by viirya**
## What changes were proposed in this pull request?

SparkSQL doesn't support to encode `Option[Product]` as a top-level row now, because in SparkSQL entire top-level row can't be null.

However for use cases like Aggregator, it is reasonable to use `Option[Product]` as buffer and output column types. Due to above limitation, we don't do it for now.

This patch proposes to encode `Option[Product]` at top-level as single struct column. So we can work around the issue that entire top-level row can't be null.

To summarize encoding of `Product` and `Option[Product]`.

For `Product`, 1. at root level, the schema is all fields are flatten it into multiple columns. The `Product ` can't be null, otherwise it throws an exception.

```scala
val df = Seq((1 -> "a"), (2 -> "b")).toDF()
df.printSchema()

root
 |-- _1: integer (nullable = false)
 |-- _2: string (nullable = true)
```

2. At non-root level, `Product` is a struct type column.

```scala
val df = Seq((1, (1 -> "a")), (2, (2 -> "b")), (3, null)).toDF()
df.printSchema()

root
 |-- _1: integer (nullable = false)
 |-- _2: struct (nullable = true)
 |    |-- _1: integer (nullable = false)
 |    |-- _2: string (nullable = true)
```

For `Option[Product]`, 1. it was not supported at root level. After this change, it is a struct type column.

```scala
val df = Seq(Some(1 -> "a"), Some(2 -> "b"), None).toDF()
df.printSchema

root
 |-- value: struct (nullable = true)
 |    |-- _1: integer (nullable = false)
 |    |-- _2: string (nullable = true)
```

2. At non-root level, it is also a struct type column.

```scala
val df = Seq((1, Some(1 -> "a")), (2, Some(2 -> "b")), (3, None)).toDF()
df.printSchema

root
 |-- _1: integer (nullable = false)
 |-- _2: struct (nullable = true)
 |    |-- _1: integer (nullable = false)
 |    |-- _2: string (nullable = true)
```

3. For use case like Aggregator, it was not supported too. After this change, we support to use `Option[Product]` as buffer/output column type.

```scala
val df = Seq(
    OptionBooleanIntData("bob", Some((true, 1))),
    OptionBooleanIntData("bob", Some((false, 2))),
    OptionBooleanIntData("bob", None)).toDF()

val group = df
    .groupBy("name")
    .agg(OptionBooleanIntAggregator("isGood").toColumn.alias("isGood"))
group.printSchema

root                                                  
 |-- name: string (nullable = true)                          
 |-- isGood: struct (nullable = true)
 |    |-- _1: boolean (nullable = false)
 |    |-- _2: integer (nullable = false)
```

The buffer and output type of `OptionBooleanIntAggregator` is both `Option[(Boolean, Int)`.

## How was this patch tested?

Added test.

### CODE REVIEW

**Comment by viirya** [comment]
retest this please.

**Comment by viirya** [comment]
retest this please.

**Comment by viirya** [comment]
I'm wondering should we add encoders of Option of Product into object `Encoders`.

**Comment by viirya** [comment]
cc @cloud-fan @hvanhovell

**Comment by cloud-fan** [comment]
how about we treat the top level `Option[Product]` as a single struct column? then we can git rid of this limitation entirely.

**Comment by viirya** [comment]
It sounds like much more behavior changing?

**Comment by cloud-fan** [comment]
yes it is, but it makes the encoder framework more consistent. And making a failure case into runnable is a safe behavior change.

**Comment by viirya** [comment]
Non top-level and top-level encoders for `Option[Product]` have a little difference.

As you said, top-level `Option[Product]` can only be encoded as a single struct column.

For non top-level one, we can't apply the same change because it is already a struct column. We don't want to change current behavior of it from a struct column to a struct column of a struct column.

This means that we can remove the limitation of top-level `Option[Product]`, but it doesn't help too much Aggregator issue here. We still need to use non top-level `Option[Product]` encoders for Aggregator case.

@cloud-fan Do you want to incorporate top-level `Option[Product]` encoders into this PR too? Or let's create another Jira & PR for it? Thanks.

**Comment by viirya** [comment]
ping @cloud-fan

**Comment by viirya** [comment]
At the end of encoder creation? You mean at the end of calling `ExpressionEncoder.apply()`? But it is used both for top-level encoder e.g., `Dataset[Option[Product]]` and  non top-level encoder e.g., `Aggregator`'s encoder. If we flatten it, doesn't it mean for top-level, it is encoded as a row, not a struct column?

File: sql/core/src/test/scala/org/apache/spark/sql/DatasetAggregatorSuite.scala [hunk 1]
```diff
@@ -183,6 +184,43 @@ case class OptionBooleanAggregator(colName: String)
   def OptionalBoolEncoder: Encoder[Option[Boolean]] = ExpressionEncoder()
 }
 
+case class OptionBooleanIntAggregator(colName: String)
+    extends Aggregator[Row, Option[(Boolean, Int)], Option[(Boolean, Int)]] {
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetAggregatorSuite.scala]: what's the expected schema after we apply an aggregator with `Option[Product]` as buffer/output?

Comment (cloud-fan) [reply sql/core/src/test/scala/org/apache/spark/sql/DatasetAggregatorSuite.scala]: assuming non top level, `Option[Product]` is same as `Product`?

Comment (viirya) [reply sql/core/src/test/scala/org/apache/spark/sql/DatasetAggregatorSuite.scala]: Yes. For non top level, `[Option[Product]` is same as `Product`. The difference is additional `WrapOption` and `UnwrapOption` around expressions.

**Comment by viirya** [comment]
ping @cloud-fan @hvanhovell

**Comment by cloud-fan** [comment]
Again, can we always support `Option[Product]` with some special handling for top-level encoder expression?

**Comment by cloud-fan** [comment]
This PR is just a special handling for `Option[Product]` in Aggregator, I think we don't need it when we have the more general solution, right?

**Comment by viirya** [comment]
retest this please.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala [hunk 2]
```diff
@@ -43,20 +43,17 @@ import org.apache.spark.util.Utils
...
  */
 object ExpressionEncoder {
-  def apply[T : TypeTag](): ExpressionEncoder[T] = {
+  // Constructs an encoder for top-level row.
+  def apply[T : TypeTag](): ExpressionEncoder[T] = apply(topLevel = true)
+
+  /**
+   * @param topLevel whether the encoders to construct are for top-level row.
+   */
+  def apply[T : TypeTag](topLevel: Boolean): ExpressionEncoder[T] = {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala]: where do we call this apply with `topLevel = false`?

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala]: In `Aggregator`, we can call this apply with `topLevel = false` to avoid resulting a nested struct.

**Comment by viirya** [comment]
retest this please.

**Comment by HyukjinKwon** [comment]
retest this please

**Comment by viirya** [comment]
retest this please.

**Comment by viirya** [comment]
retest this please.

**Comment by viirya** [comment]
retest this please.

**Comment by viirya** [comment]
retest this please.

**Comment by cloud-fan** [comment]
Can you give a description of the new encoding rules regarding `Option`?

e.g.
```
Option[Int] in normal encoder -> a nullable int column
Option[Int] in agg encoder -> a nullable int column
Option[Int] in tupled encoder -> a nullable int column

Option[Product] in normal encoder => a nullable struct column
Option[Product] in agg encoder => ?
...
```

**Comment by viirya** [comment]
Yeah. The encoding rules regarding `Option` are:

1. For non `Product` types, no change is made.
```
Option[Int] in normal encoder -> a nullable int column
Option[Int] in agg encoder -> a nullable int column
Option[Int] in tupled encoder -> a nullable int column
```

E.g.,

```scala
val df = Seq(Some(1), Some(2), None).toDF()
df.printSchema

root
 |-- value: integer (nullable = true)

val df2 = Seq((1, Some(1)), (2, Some(2)), (3, None)).toDF()
df2.printSchema

root
 |-- _1: integer (nullable = false)
 |-- _2: integer (nullable = true)

```


2. For `Product` types, we don't support it at top level and aggregator for now. After this change, we can support it in normal encoder and aggregator.

```
Option[Product] in normal encoder => a nullable struct column
```

E.g.,

```scala
val df3 = Seq(Some(1 -> "a"), Some(2 -> "b"), None).toDF()
df3.printSchema

root
 |-- value: struct (nullable = true)
 |    |-- _1: integer (nullable = false)
 |    |-- _2: string (nullable = true)
```

```
Option[Product] in agg encoder => a nullable struct column // topLevel = false
Option[Product] in agg encoder => a nullable struct column nested in a struct column // topLevel = true
```

```scala
val df4 = Seq(
    OptionBooleanIntData("bob", Some((true, 1))),
    OptionBooleanIntData("bob", Some((false, 2))),
    OptionBooleanIntData("bob", None)).toDF()

val group = df4
    .groupBy("name")
    .agg(OptionBooleanIntAggregator("isGood").toColumn.alias("isGood"))
group.printSchema

// topLevel = false
root                                                  
 |-- name: string (nullable = true)                          
 |-- isGood: struct (nullable = true)
 |    |-- _1: boolean (nullable = false)
 |    |-- _2: integer (nullable = false)

// topLevel = true
root                    
 |-- name: string (nullable = true)                   
 |-- isGood: struct (nullable = true)                        
 |    |-- value: struct (nullable = true)
 |    |    |-- _1: boolean (nullable = false)
 |    |    |-- _2: integer (nullable = false)
```

3. For `Option[Product]` in tupled encoder, no change is made too:

```scala
val df5 = Seq((1, Some(1 -> "a")), (2, Some(2 -> "b")), (3, None)).toDF()
df5.printSchema

root
 |-- _1: integer (nullable = false)
 |-- _2: struct (nullable = true)
 |    |-- _1: integer (nullable = false)
 |    |-- _2: string (nullable = true)

```

**Comment by viirya** [comment]
From the above, I think the aggregator encoder for `Option[Product]` might be a bit tricky to use for users, since they might need to know the difference between `topLevel = true` and `topLevel = false`. But I don't have good idea for now on this. @cloud-fan WDYT?

**Comment by cloud-fan** [comment]
if we can totally hide the `topLevel` parameter from users, it should be fine. Generally I think the behavior is consistent, now `Option[Product]` is always a struct type column. But we need to improve the implementation to hide the `topLevel`.

**Comment by cloud-fan** [comment]
The only tricky thing is, `Product` is handled specially in the top level, being flattened into multiple columns.

**Comment by viirya** [comment]
> if we can totally hide the topLevel parameter from users, it should be fine. Generally I think the behavior is consistent, now Option[Product] is always a struct type column. But we need to improve the implementation to hide the topLevel.

Sounds good. I'm wondering if there is good way to hide `topLevel`.

**Comment by viirya** [comment]
@cloud-fan I made attempt to remove `topLevel` parameter. The approach is to flatten serializers and deserialzer at `TypedAggregateExpression`. So users are not aware of difference when using expression encoder.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TypedAggregateExpression.scala [hunk 3]
```diff
@@ -19,25 +19,85 @@ package org.apache.spark.sql.execution.aggregate
...
+  def isOptProductEncoder(encoder: ExpressionEncoder[_]): Boolean = {
+    // Only Option[Product] is non-flat.
+    encoder.clsTag.runtimeClass == classOf[Option[_]] && !encoder.flat
+  }
+
+  /**
+   * Flattens serializers and deserializer of given encoder. We only flatten encoder
+   * of `Option[Product]` class.
+   */
+  def flattenOptProductEncoder(encoder: ExpressionEncoder[_]): ExpressionEncoder[_] = {
```
Comment (viirya) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TypedAggregateExpression.scala]: I will go to add some tests against this.

**Comment by viirya** [comment]
retest this please.

**Comment by viirya** [comment]
retest this please.

**Comment by viirya** [comment]
retest this please.

**Comment by viirya** [comment]
> The only tricky thing is, Product is handled specially in the top level, being flattened into multiple columns.

@cloud-fan Compared with Option of Product which is not supported before, the encoding of Product is current behavior. I think we don't need to change it so far. WDYT?

**Comment by dilipbiswal** [comment]
retest this please

**Comment by viirya** [comment]
retest this please.

**Comment by cloud-fan** [comment]
Since we have to handle some special cases, can you list them in the PR description?

For `Product`, normally it should be a struct type column, the special case I can think of is: 1) at root level, flatten it into multiple columns, and the `Product` can't be null.

For `Option[Product]`, normally it should also be a struct type column, seems there is no special case.

**Comment by viirya** [comment]
> For Product, normally it should be a struct type column, the special case I can think of is: 1) at root level, flatten it into multiple columns, and the Product can't be null.
> For Option[Product], normally it should also be a struct type column, seems there is no special case.

This doesn't make any change to `Product`.

For `Product`, 1. at root level, the schema is all fields are flatten it into multiple columns. The Product can't be null.

```scala
val df = Seq((1 -> "a"), (2 -> "b")).toDF()
df.printSchema()

root
 |-- _1: integer (nullable = false)
 |-- _2: string (nullable = true)
```

2. At non-root level, `Product` is a struct type column.

```scala
val df = Seq((1, (1 -> "a")), (2, (2 -> "b")), (3, null)).toDF()
df.printSchema()

root
 |-- _1: integer (nullable = false)
 |-- _2: struct (nullable = true)
 |    |-- _1: integer (nullable = false)
 |    |-- _2: string (nullable = true)
```

For `Option[Product]`, I've posted some description in the PR description.

**Comment by cloud-fan** [comment]
> This doesn't make any change to Product.

yea I know, but we should put things together to evaluate a design.

After reading your description, `Option[Product]` is the same as `Product` except at top level?

**Comment by viirya** [comment]
> Option[Product] is the same as Product except at top level?

Yeah, at top level, `Option[Product]` was not supported. After this change, it is a struct type column.

**Comment by viirya** [comment]
Thanks. I've updated the PR description to summarize encoding of `Product` and `Option[Product]`.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala [hunk 4]
```diff
@@ -135,19 +135,62 @@ object ScalaReflection extends ScalaReflection {
...
    */
-  def deserializerFor[T : TypeTag]: Expression = {
+  def deserializerFor[T : TypeTag]: Expression = cleanUpReflectionObjects {
     val tpe = localTypeOf[T]
     val clsName = getClassNameFromType(tpe)
     val walkedTypePath = s"""- root class: "$clsName"""" :: Nil
-    val expr = deserializerFor(tpe, None, walkedTypePath)
-    val Schema(_, nullable) = schemaFor(tpe)
+    val Schema(dataType, tpeNullable) = schemaFor(tpe)
+    val isOptionOfProduct = tpe.dealias <:< localTypeOf[Option[_]] &&
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/ScalaReflection.scala]: can we reuse the `optionOfProductType`?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/ScalaReflectionSuite.scala [hunk 5]
```diff
@@ -358,4 +358,31 @@ class ScalaReflectionSuite extends SparkFunSuite {
     assert(numberOfCheckedArguments(deserializerFor[(java.lang.Double, Int)]) == 1)
     assert(numberOfCheckedArguments(deserializerFor[(java.lang.Integer, java.lang.Integer)]) == 0)
   }
+
+  test("SPARK-24762: serializer for Option of Product") {
+    val optionOfProduct = Some((1, "a"))
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/ScalaReflectionSuite.scala]: why do we need this? we can write `classOf[Option[(Int, String)]]` below.

**Comment by viirya** [comment]
Rebased this after refactoring ExpressionEncoder.

File: docs/sql-migration-guide-upgrade.md [hunk 6]
```diff
@@ -17,6 +17,8 @@ displayTitle: Spark SQL Upgrading Guide
 
   - The `ADD JAR` command previously returned a result set with the single value 0. It now returns an empty result set.
 
+  - In Spark version 2.4 and earlier, `Dataset` doesn't support to encode `Option[Product]` at top-level row, because in Spark SQL entire top-level row can't be null. Since Spark 3.0, `Option[Product]` at top-level is encoded as a row with single struct column. Then with this support, `Aggregator` can also use use `Option[Product]` as buffer and output column types.
```
Comment (cloud-fan) [inline docs/sql-migration-guide-upgrade.md]: Usually we only add migration guide if something is broken and users must be aware of it when upgrading.

I think this one is not the case? It's just a new feature.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala [hunk 7]
```diff
@@ -207,7 +198,7 @@ case class ExpressionEncoder[T](
   val serializer: Seq[NamedExpression] = {
     val clsName = Utils.getSimpleName(clsTag.runtimeClass)
 
-    if (isSerializedAsStruct) {
+    if (isSerializedAsStruct && !classOf[Option[_]].isAssignableFrom(clsTag.runtimeClass)) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala]: can we make sure that, other places calling `isSerializedAsStruct` don't need to check Option type?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/ScalaReflectionSuite.scala [hunk 8]
```diff
@@ -362,4 +362,38 @@ class ScalaReflectionSuite extends SparkFunSuite {
     assert(numberOfCheckedArguments(deserializerFor[(java.lang.Double, Int)]) == 1)
     assert(numberOfCheckedArguments(deserializerFor[(java.lang.Integer, java.lang.Integer)]) == 0)
   }
+
+  test("SPARK-24762: serializer for Option of Product") {
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/ScalaReflectionSuite.scala]: do we need to add tests here? `ScalaReflection` is not updated in this PR.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TypedAggregateExpression.scala [hunk 9]
```diff
@@ -76,7 +76,7 @@ object TypedAggregateExpression {
         None,
         bufferSerializer,
         bufferEncoder.resolveAndBind().deserializer,
-        outputEncoder.serializer,
+        outputEncoder.objSerializer,
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TypedAggregateExpression.scala]: to confirm, this is a un-related change and just clean up the code?

Comment (viirya) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/TypedAggregateExpression.scala]: This is required. Without this, the output schema of aggregator using Option[Product] as output encoder is not correct.

File: docs/sql-migration-guide-upgrade.md (hunk 6 above)
Comment (viirya) [reply docs/sql-migration-guide-upgrade.md]: Ok. Let me revert this.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/ScalaReflectionSuite.scala (hunk 8 above)
Comment (viirya) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/ScalaReflectionSuite.scala]: It is mainly because I don't find related tests for serializer and deserializer of Option[Product]. I can remove it or make a separate PR for it, since it is not related to the change in this PR.

File: sql/core/src/test/scala/org/apache/spark/sql/DatasetAggregatorSuite.scala [hunk 10]
```diff
@@ -393,4 +431,18 @@ class DatasetAggregatorSuite extends QueryTest with SharedSQLContext {
...
+  test("SPARK-24762: Aggregator should be able to use Option of Product encoder") {
+    val df = Seq(
+      OptionBooleanIntData("bob", Some((true, 1))),
+      OptionBooleanIntData("bob", Some((false, 2))),
+      OptionBooleanIntData("bob", None)).toDF()
+
+    val group = df
+      .groupBy("name")
+      .agg(OptionBooleanIntAggregator("isGood").toColumn.alias("isGood"))
+    assert(df.schema == group.schema)
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetAggregatorSuite.scala]: let's write down the expected schema
```
val expectedSchema = ...
assert(df.schema == expectedSchema)
assert(grouped.schema == ...)
```

File: sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala [hunk 11]
```diff
@@ -1556,6 +1547,54 @@ class DatasetSuite extends QueryTest with SharedSQLContext {
...
+
+  test("SPARK-24762: Enable top-level Option of Product encoders") {
+    val data = Seq(Some((1, "a")), Some((2, "b")), None)
+    val ds = data.toDS()
+
+    checkDataset(
+      ds,
+      data: _*)
+
+    val schema = StructType(Seq(
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala]: can we use the `add` API? e.g.
```
new StructType().add(
  "value",
  new StructType()
    .add("_1", ...)
    .add("_2", ...))
```

**Comment by cloud-fan** [comment]
I think this is close, can you answer https://github.com/apache/spark/pull/21732/files#r228782670 ?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala (hunk 7 above)
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala]: I think some places are needed to check Option too. I will also add few more tests to cover some use cases.

One possible place is Dataset.groupByKey. Before that, I may need #22944 to be merged first. So I can write something like:

```scala
val ds = Seq(Some(("a", 1)), Some(("b", 2)), Some(("c", 3))).toDS()
ds.groupByKey(_.map(_._2).getOrElse("d")).agg(sum("value._2").as[Long], sum($"value._2" + 1).as[Long])
```

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala]: I add more necessary Option type check in few places. And also related tests.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala [hunk 12]
```diff
@@ -198,7 +189,7 @@ case class ExpressionEncoder[T](
   val serializer: Seq[NamedExpression] = {
     val clsName = Utils.getSimpleName(clsTag.runtimeClass)
 
-    if (isSerializedAsStruct) {
+    if (isSerializedAsStruct && !isOptionType) {
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala]: if we always check them together, how about putting `!isOptionType` inside `isSerializedAsStruct`?

File: sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala [hunk 13]
```diff
@@ -1556,6 +1547,69 @@ class DatasetSuite extends QueryTest with SharedSQLContext {
...
+          StructField("_1", IntegerType, nullable = false),
+          StructField("_2", StringType, nullable = true)))),
+        StructField("_2", DoubleType, nullable = false)
+      )), nullable = true)
+    ))
+    assert(nestedDs.schema == nestedSchema)
+  }
+
+  test("SPARK-24762: Resolving Option[Product] field") {
+    val ds = Seq((1, ("a", 1.0)), (2, ("b", 2.0))).toDS().as[(Int, Option[(String, Double)])]
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala]: we can put a null value in the input data?

File: sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala [hunk 14]
```diff
@@ -1556,6 +1547,69 @@ class DatasetSuite extends QueryTest with SharedSQLContext {
...
+    checkDataset(ds,
+      (1, Some(("a", 1.0))), (2, Some(("b", 2.0))))
+  }
+
+  test("SPARK-24762: select Option[Product] field") {
+    val ds = Seq(("a", 1), ("b", 2), ("c", 3)).toDS()
+      .select(expr("struct(_2, _2 + 1)").as[Option[(Int, Int)]])
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala]: can we also test null values here? e.g. `if(_2 > 2, struct..., null)`

File: sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala [hunk 15]
```diff
@@ -1556,6 +1547,69 @@ class DatasetSuite extends QueryTest with SharedSQLContext {
...
+    checkDataset(ds,
+      Some((1, 2)), Some((2, 3)), Some((3, 4)))
+  }
+
+  test("SPARK-24762: joinWith on Option[Product]") {
+    val ds1 = Seq(Some((1, 2)), Some((2, 3))).toDS().as("a")
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/DatasetSuite.scala]: ditto, let's test None

**Comment by cloud-fan** [comment]
LGTM except a few comments, good job!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala (hunk 12 above)
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala]: There is a `isSerializedAsStruct` check in `TypedAggregateExpression` which doesn't come with `isOptionType`.

I'd think `isSerializedAsStruct` is true or false only depending how we serialize object `T` and isn't related to how we use (flatten or not) it in `ExpressionEncoder`. At least it is more clear to me.

**Comment by viirya** [comment]
retest this please.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala [hunk 16]
```diff
@@ -257,6 +251,11 @@ case class ExpressionEncoder[T](
    */
   def isSerializedAsStruct: Boolean = objSerializer.dataType.isInstanceOf[StructType]
 
+  /**
+   * Returns true if the type `T` is `Option`.
+   */
+  def isOptionType: Boolean = classOf[Option[_]].isAssignableFrom(clsTag.runtimeClass)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala]: regarding maintainability, I think it's better to add a 
```
// maybe there is a better naming
def isSerializedAsStructForTopLevel: Boolean = {
  isSerializedAsStruct && isOption
}
```

The benefit is, when people call `isSerializedAsStruct`, they will see this similar method and think carefully which one he should use. Otherwise, people may miss to check `isOptionType` easily.

**Comment by cloud-fan** [comment]
last comment, LGTM otherwise

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala (hunk 16 above)
Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala]: How do we use `isSerializedAsStructForTopLevel`? In all cases now, we only need to check `isSerializedAsStruct && !isOptionType`.

**Comment by cloud-fan** [comment]
thanks, merging to master, great work!

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala [hunk 17]
```diff
@@ -253,10 +247,24 @@ case class ExpressionEncoder[T](
...
+
+  /**
+   * If the type `T` is serialized as a struct, when it is encoded to a Spark SQL row, fields in
+   * the struct are naturally mapped to top-level columns in a row. In other words, the serialized
+   * struct is flattened to row. But in case of the `T` is also an `Option` type, it can't be
+   * flattened to top-level row, because in Spark SQL top-level row can't be null. This method
+   * returns true if `T` is serialized as struct and is not `Option` type.
+   */
+  def isSerializedAsStructForTopLevel: Boolean = isSerializedAsStruct && !isOptionType
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala]: can you send a followup PR to inline `isOptionType` if it's only used here?

Comment (viirya) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/encoders/ExpressionEncoder.scala]: ok.

**Comment by HyukjinKwon** [comment]
Hm .. sorry for joining this party late. I was reading and testing it by myself.


```scala
scala> Seq((1, "a"), (2, "b")).toDF.show()
+---+---+
| _1| _2|
+---+---+
|  1|  a|
|  2|  b|
+---+---+


scala> Seq(1, 2).toDF.show()
+-----+
|value|
+-----+
|    1|
|    2|
+-----+
```


```scala
scala> Seq(Some((1, "a")), Some((2, "b"))).toDF.show()
+------+
| value|
+------+
|[1, a]|
|[2, b]|
+------+


scala> Seq(Some(1), Some(2)).toDF.show()
+-----+
|value|
+-----+
|    1|
|    2|
+-----+
```

I think this behaviour can be actually controversial. If we interpret `Option` as existent/missing value. Why did we interpret `Option` as `Tuple1`?

**Comment by cloud-fan** [comment]
`Seq(Some(1), Some(2))` is treated same as `Seq(1, 2)`, because it's only a single column and we can make it nullable. So the `Some` here changes nothing but the column nullability.

`Seq((1, "a"),(2, "b"))` is special, as dataset encoder assumes the top level `Product` will never be null, and flatten it into 2 columns. If we add `Some` here, we can't do the flattening.

**Comment by HyukjinKwon** [comment]
re: https://github.com/apache/spark/pull/21732#issuecomment-450160499 I was thinking both below

```scala
Seq(Some((1, "a")), Some((2, "b"))).toDF.show()
Seq((1, "a"), (2, "b")).toDF.show()
```

should produce the same result since

```scala
Seq(Some(1), Some(2)).toDF.show()
Seq(1, 2).toDF.show()
```

produces the same results anyhow. Apparently this is why it has been disallowed.

**Comment by viirya** [comment]
Thanks for @cloud-fan's explanation. So I think @HyukjinKwon you mean why we interpret `Option[Product]` like `Some((1, "a")), Some((2, "b"))` as `Tuple2`.

For the following
```
Seq(Some((1, "a")), Some((2, "b"))).toDF.show()
Seq((1, "a"), (2, "b")).toDF.show()
```

can't produce the same result since top-level null row is not allowed in Spark.

**Comment by HyukjinKwon** [comment]
Yea, then why did we allow the different result? I was thinking we're going to allow this for aggregators.

**Comment by viirya** [comment]
@HyukjinKwon there was a comment https://github.com/apache/spark/pull/21732#issuecomment-403865941 for it. This was originally to make `Option[Product]` work for aggregators, but it makes encoder framework inconsistent because the created encoders can used for aggregators but not for Dataset. So I think the idea is to make it consistent. Users are allowed to encoder `Option[Product]` as Dataset but they should be aware of the encoding.

**Comment by HyukjinKwon** [comment]
Hm, for aggregators, I would consider this as non root level. Looks they use the same encoder but can't be the same.

**Comment by cloud-fan** [comment]
@HyukjinKwon If we can go back, I'd say we should not have this optimization which flattens top-level `Product` when encoding. This brings the assumption that top level `Product` can't be null, because in Spark the top level `Row` can't be null.

Ideally `Option[T]` should be the same as `T` if `T` is nullable. It's just 2 different ways to represent null in Scala and Java. But because of the optimization I mentioned before, top-level `Product` is the only exception.

It's too late to revert that optimization, I think we should accept this special case.

**Comment by HyukjinKwon** [comment]
I didn't mean that we should revert .. was just checking the PRs in my queue and was just curious.

I mean, I understood the limitation but failed to understand  why it's been allowed. We exposed `Option[T < Project]` at top level but limitation is still there.

**Comment by HyukjinKwon** [comment]
Ah, but you're saying `Product` at top-level is an exception and the other cases are all coherent? hmm ... okie.

**Comment by cloud-fan** [comment]
Yes, top-level `Product` is the only exception, because of the flatten trick which has been there for years.
